src/k8s/api/v1/types.go

```diff
@@ -119,20 +119,74 @@ func (c ClusterStatus) String() string {
 	} else {
 		result.WriteString("no")
 	}
-	result.WriteString("\n\n")
-	result.WriteString("control-plane nodes:\n")
-	for _, member := range c.Members {
-		// There is not much that we can do if the hostport is wrong.
-		// Thus, ignore the error and just display an empty IP field.
-		apiServerIp, _, _ := net.SplitHostPort(member.Address)
-		result.WriteString(fmt.Sprintf("  %s: %s\n", member.Name, apiServerIp))
+	result.WriteString("\n")
+	result.WriteString("datastore:\n")
+
+	voters := make([]NodeStatus, 0, len(c.Members))
+	standBys := make([]NodeStatus, 0, len(c.Members))
+	spares := make([]NodeStatus, 0, len(c.Members))
+	for _, node := range c.Members {
+		switch node.DatastoreRole {
+		case DatastoreRoleVoter:
+			voters = append(voters, node)
+		case DatastoreRoleStandBy:
+			standBys = append(standBys, node)
+		case DatastoreRoleSpare:
+			spares = append(spares, node)
+		}
+	}
+	if len(voters) > 0 {
+		result.WriteString(fmt.Sprintf("  master-nodes:\n"))
+		for _, voter := range voters {
+			result.WriteString(fmt.Sprintf("    - %s\n", voter.Address))
+		}
+	} else {
+		result.WriteString(fmt.Sprintf("  master-nodes: none\n"))
+	}
+	if len(standBys) > 0 {
+		result.WriteString(fmt.Sprintf("  standy-nodes:\n"))
+		for _, standBy := range standBys {
+			result.WriteString(fmt.Sprintf("    - %s\n", standBy.Address))
+		}
+	} else {
+		result.WriteString(fmt.Sprintf("  standy-nodes: none\n"))
+	}
+	if len(spares) > 0 {
+		result.WriteString(fmt.Sprintf("  spare-nodes:\n"))
+		for _, spare := range spares {
+			result.WriteString(fmt.Sprintf("    - %s\n", spare.Address))
+		}
+	} else {
+		result.WriteString(fmt.Sprintf("  spare-nodes: none\n"))
 	}
 	result.WriteString("\n")
 
-	result.WriteString("components:\n")
-	for _, component := range c.Components {
-		result.WriteString(fmt.Sprintf("  %-10s %s\n", component.Name, component.Status))
+	printedConfig := UserFacingClusterConfig{}
+	if c.Config.Network.Enabled != nil && *c.Config.Network.Enabled {
+		printedConfig.Network = c.Config.Network
+	}
+	if c.Config.DNS.Enabled != nil && *c.Config.DNS.Enabled {
+		printedConfig.DNS = c.Config.DNS
+	}
+	if c.Config.Ingress.Enabled != nil && *c.Config.Ingress.Enabled {
+		printedConfig.Ingress = c.Config.Ingress
+	}
+	if c.Config.LoadBalancer.Enabled != nil && *c.Config.LoadBalancer.Enabled {
+		printedConfig.LoadBalancer = c.Config.LoadBalancer
 	}
+	if c.Config.LocalStorage.Enabled != nil && *c.Config.LocalStorage.Enabled {
+		printedConfig.LocalStorage = c.Config.LocalStorage
+	}
+	if c.Config.Gateway.Enabled != nil && *c.Config.Gateway.Enabled {
+		printedConfig.Gateway = c.Config.Gateway
```

Comment:
How about something like this?

```suggestion
		updatePrintedConfigIfEnabled(c.Config.Network, func() {
             printedConfig.Network = c.Config.Network
        })
        updatePrintedConfigIfEnabled(c.Config.DNS, func() {
             printedConfig.DNS = c.Config.DNS
        })
```

Reply:
hm, the `updatePrintedConfigIfEnabled` would just abstract away the nil check, right?
I don't think this is more readable or shorter than the if statement, is it?

Reply:
That is true